### PR TITLE
feat: standalone testing check

### PR DESCRIPTION
**Pull request by ryan-timothy-albert**
Moves forward with us running testing by default as a PR check
- Allows the user to attach their own ACTION_PAT to automatically trigger tests
- Also leaves the door open to use using our app to trigger tests either with empty commits or opening/closing PRs

### CODE REVIEW

**Comment by ryan-timothy-albert** [comment]
This actual auto generated github file would look like the following

```
name: Test [SDK Name]
permissions:
  checks: write
  contents: write
  pull-requests: write
  statuses: write
"on":
  workflow_dispatch:
    inputs:
      target:
        description: Specific target to test
        type: string
  schedule:
    - cron: 0 0 * * *
jobs:
  generate:
    uses: speakeasy-api/sdk-generation-action/.github/workflows/sdk-test.yaml@v15
    with:
      target: ${{ github.event.inputs.force }}
      working_directory: 'nested'
    secrets:
      github_access_token: ${{ secrets.GITHUB_TOKEN }}
      speakeasy_api_key: ${{ secrets.SPEAKEASY_API_KEY }}
```

**Comment by TristanSpeakEasy** [comment]
LGTM but I will let @chase-crumbaugh approve

File: .github/workflows/sdk-test.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    steps:
+      - name: Tune GitHub-hosted runner network
+        uses: smorimoto/tune-github-hosted-runner-network@v1
+
+      - name: Check commit message condition
+        id: check_commit
+        run: |
+          if [[ "${{ github.event_name }}" != "push" ]]; then
+            echo "Skipping commit message check since event is not push."
+            echo "run_tests=true" >> $GITHUB_OUTPUT
```
Comment (chase-crumbaugh) [inline .github/workflows/sdk-test.yaml:34]: Should this be false?

File: .github/workflows/sdk-test.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            exit 0
+          fi
+
+          COMMIT_MESSAGE="${{ github.event.head_commit.message }}"
+          echo "Commit message: $COMMIT_MESSAGE"
+
+          # The main trigger should be PR events. This allows us to execute tests from empty commits made by our app.
+          # This is a workaround for GitHub not allowing actions to trigger from a PR made by another action.
+          if [[ "$COMMIT_MESSAGE" == *"[run-tests]"* ]]; then
```
Comment (chase-crumbaugh) [inline .github/workflows/sdk-test.yaml:43]: Slightly confused by the relationship between this `run-tests` and the `run_tests` you are writing to the output

File: .github/workflows/workflow-executor.yaml [hunk 3]
```diff
@@ -109,6 +109,9 @@ on:
       speakeasy_api_key:
         description: The API key to use to authenticate the Speakeasy CLI
         required: true
+      action_pat:
+        description: A specific Github PAT used to create Pull Requests
+        required: false
```
Comment (chase-crumbaugh) [inline .github/workflows/workflow-executor.yaml:114]: Wouldn't they just provide this in `github_access_token` ? We should be able to tell what type of token it is from the token itself

File: internal/actions/test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	if _, err = cli.Download("latest", g); err != nil {
+		return err
+	}
+
+	wf, err := configuration.GetWorkflowAndValidateLanguages(false)
+	if err != nil {
+		return err
+	}
+
+	// This will only come in via workflow dispatch, we do accept 'all' as a validate special case
```
Comment (chase-crumbaugh) [inline internal/actions/test.go:31]: nit: typo: validate

File: internal/actions/test.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	var testedTargets []string
+	if providedTargetName := environment.SpecifiedTarget(); providedTargetName != "" {
+		testedTargets = append(testedTargets, providedTargetName)
+	}
+
+	if len(testedTargets) == 0 {
+		files, err := g.GetChangedFilesForPR()
```
Comment (chase-crumbaugh) [inline internal/actions/test.go:38]: Maybe worth explaining in a comment what we are doing here and why it's necessary

File: .github/workflows/sdk-test.yaml (hunk 2 above)
Comment (ryan-timothy-albert) [reply .github/workflows/sdk-test.yaml:43]: It's a special commit message for our empty commit flow

Comment (ryan-timothy-albert) [reply .github/workflows/sdk-test.yaml:43]: Basically avoids us having duplicate runs with the PR trigger. If we make an empty commit with our app from our backend it will have [run-tests] and thus trigger tests.

File: .github/workflows/sdk-test.yaml (hunk 1 above)
Comment (ryan-timothy-albert) [reply .github/workflows/sdk-test.yaml:34]: No this is a very specific check for PUSH events

Comment (ryan-timothy-albert) [reply .github/workflows/sdk-test.yaml:34]: If it's a PUSH event we want to avoid duplicate runs. We only have this to support our somewhat hacky app based empty commit which will have that special message

File: .github/workflows/workflow-executor.yaml (hunk 3 above)
Comment (ryan-timothy-albert) [reply .github/workflows/workflow-executor.yaml:114]: Because they will need to setup way more permissions then. Replacing GITHUB_TOKEN entirely is unnecessary and just makes it way more likely something gets messed up in the set up process.

File: internal/actions/test.go (hunk 5 above)
Comment (ryan-timothy-albert) [reply internal/actions/test.go:38]: will do

File: internal/actions/test.go (hunk 4 above)
Comment (ryan-timothy-albert) [reply internal/actions/test.go:31]: will do

File: .github/workflows/sdk-test.yaml (hunk 2 above)
Comment (ryan-timothy-albert) [reply .github/workflows/sdk-test.yaml:43]: I'll try to explain better in the code comment. That was my intention.

**Review by chase-crumbaugh — APPROVED** [review]
